### ISSUE TITLE
setup dev env buckets

### ISSUE THREAD

**Issue by drstarson** [Passiolife/passio_devops#4]
- [x] @Loenike setup dev env buckets to enable the full flow of images we have now 
- [x] add documentation as need

**Comment by Loenike** [comment]
I added the following 6 buckets to the dev-env project:

```
passio-dev-env-assets
passio-dev-env-assets-original
passio-dev-env-assets-original-direct
passio-dev-env-cropped-assets
passio-dev-env-staging
passio-dev-env-staging-direct
```

With these buckets, we should be able to setup the following processes:
a) in the **passio-dev-env-staging** bucket we have folders - named by **passioID_sourceID**
b) we should create a process (cloud function?) to upload those blobs in the **passio-dev-env-assets-original** and use the updated API to push them to the assets service and update also passioID and sourceID in the database. after they have been processed by the assets service, the new assets will be in the bucket **passio-dev-env-assets**
c) we can setup a parallel process to add images to the database and SKIP the inspector queue - for those I added the buckets **passio-dev-env-staging-direct** - and - **passio-dev-env-assets-original-direct**
d) once an image has been processed by the queue / added to the database - the cropping process should take place - those cropped assets will be stored in **passio-dev-env-cropped-assets**

This means:
- our actual image assets will be stored in **passio-dev-env-assets** (entire image) and **passio-dev-env-cropped-assets** (cropped images)
- the buckets **passio-dev-env-assets-original** and **passio-dev-env-assets-original-direct** are the buckets where we include a "trigger" for pub sub
- **passio-dev-env-staging** and **passio-dev-env-staging-direct** are the staging buckets where we store data already in "correct" format, before we start the inclusion in the database


I added this information to the architecture visualization: https://drive.google.com/file/d/1O-B7AprLraZCO-Y9w18juEJC_j2R8Fkh/view?usp=sharing

There might be additional "hidden" buckets required for the assets service to run consistently with the current setup, which I don't know of - this might be something we could double-check with Tim.

**Comment by drstarson** [comment]
@Loenike @sptrakesh At a later stage we may also add two backup buckets **passio-dev-env-assets-backup** and **passio-dev-env-cropped-assets-backup** and do regular snapshots of **passio-dev-env-assets** and **passio-dev-env-cropped-assets**

- [ ] If this makes sense we may add the two empty buckets now. @Loenike @sptrakesh your call.

**Comment by drstarson** [comment]
@Loenike 

It sounds like we need to think through and add consistency checks to make sure assets and cropped-assets are synced. For example, if we change a few tags for cropped images via Bulk Inspector we need to propagate those changes into the assets and reflect them in the non-cropped images. 

Similarly, if we are changing a bbox or deleting bbox via the main Inspector, we need to propagate those changes into the cropped assets. 

- [ ] @Loenike these are just comments and observations. Please drive the creation of relevant issues.

**Comment by drstarson** [comment]
@Loenike I added a quick comment to the draw.io diagram

<img width="791" alt="Screen Shot 2020-03-23 at 3 56 00 PM" src="https://user-images.githubusercontent.com/17357591/77370804-e3695900-6d1e-11ea-9213-679ba9696ee8.png">

**Comment by drstarson** [comment]
@sptrakesh @Loenike 

> we should create a process (cloud function?) to upload all data (==images) from the staging bucket into the passio-dev-env-assets-original bucket. 

This is really great! I think this leads us to immediate issue that Rakesh can help with. I suggest we start creating specific GitHub based on the discussion above and assigning them.  The one above lands itself nicely to a new issue.

**Comment by drstarson** [comment]
I transferred Leonie's write up to https://github.com/Passiolife/passio_devops/wiki/Data-flow:-Buckets 
Overall this looks really great and adds a lot of clarity.

**Comment by sptrakesh** [comment]
> @Loenike @sptrakesh At a later stage we may also add two backup buckets **passio-dev-env-assets-backup** and **passio-dev-env-cropped-assets-backup** and do regular snapshots of **passio-dev-env-assets** and **passio-dev-env-cropped-assets**

@drstarson Is the purpose true disaster recovery, or just to ensure availability?  Buckets are multi-az by default, so there should be multiple copies of the asset across the underlying storage system [locations](https://cloud.google.com/storage/docs/locations).

**Comment by drstarson** [comment]
@sptrakesh Great point. I wanted to see your comments on this. The purpose is data safety and making sure we have backups. I generally would like to have a regular (say weekly or monthly) backup of the database & assets as distinct artifacts, but I do realize that GCP provides multiple layers of redundancy. The backup is probably not an urgent issue for now.

**Comment by sptrakesh** [comment]
@drstarson on the subject of database backups, I noticed that our current set up is not HA.  When we stand up our new `prod`, we should ensure the database is set up with replica.  Not sure how GCP handles failover, but general strategy if *master* fails is to promote a *slave* as the new *master*, just reducing down time.  PostgreSQL does not have auto failover, but I assume GCP has standard recipes for that, once HA cluster is set up.

Database backup is still required, and there should be automatic settings we can use to enable periodic snapshots.  For storage buckets, there is *Storage Transfer Service*, which we can configure to backup the necessary buckets.

**Comment by drstarson** [comment]
@sptrakesh Really great comments! I created two relevant issues to address the above.

**Comment by Loenike** [comment]
I added my comments to the separate issues + created a new one for the consistency of the cropped-assets bucket and the separate queue for direct upload.

> @Loenike I added a quick comment to the draw.io diagram

I believe, it would be best to keep them separate. This aligned with the setup for the "normal" upload and gives us more flexibility in terms of scheduling uploads in the future or having a staging area where we could potentially modify things WITHOUT yet causing any direct changes in the database itself

**Comment by drstarson** [comment]
@Loenike great! 

The core of this discussion is captured in the wiki and the diagram.

- [ ] @Loenike @sptrakesh Looks like this issue is ready to be closed. Please re-open if I missed something.
